File: ios/Sources/KlinikAPI/SurveysAPI.swift
import Foundation
import KlinikCore

/**
 * Patient-reported outcome questionnaires (spec M18, T6.7).
 *
 * The patient's screen and the clinician's screen read the same data and must
 * not say the same things about it. A patient sees the questions and their own
 * answers; they never see a finding, because "your reported pain has worsened"
 * is a clinical reading and this is not the thing that should deliver one.
 */

public enum SurveyAnswerType: String, Decodable, Sendable, Equatable {
    case scale0to10 = "SCALE_0_10"
    case yesNo = "YES_NO"
    case text = "TEXT"
}

public enum SurveyDirection: String, Decodable, Sendable, Equatable {
    /// Pain, swelling: a higher answer is a worse one.
    case higherIsWorse = "higher-is-worse"
    /// Sleep, satisfaction.
    case higherIsBetter = "higher-is-better"
}

public struct SurveyQuestion: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    public let text: String
    public let type: SurveyAnswerType
    /// Which way is bad. Absent on questions that are not scales.
    public let direction: SurveyDirection?
    public let alarmAt: Int?
    public let required: Bool?

    public var isRequired: Bool { required == true }

    /// Whether a chart can plot this one.
    public var isNumeric: Bool { type == .scale0to10 || type == .yesNo }
}

public struct PendingSurvey: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    public let title: String
    public let description: String?
    /// Days after the operation this one is about.
    public let milestoneDays: Int
    public let scheduledFor: Date
    /// After this it can no longer be answered.
    public let expiresAt: Date?
    public let questions: [SurveyQuestion]

    /**
     * Whether it is still open.
     *
     * A late answer is refused by the server, so the form must not invite one:
     * asking somebody to fill in five questions and then rejecting them is
     * worse than not asking.
     */
    public func isOpen(at moment: Date = Date()) -> Bool {
        guard let expiresAt else { return true }

        return expiresAt > moment
    }

    public var requiredQuestions: [SurveyQuestion] { questions.filter(\.isRequired) }
}

/// What the patient is told back. Deliberately not a reading of their answers.
public struct SurveySubmitResult: Decodable, Sendable, Equatable {
    /// True when a very satisfied patient was invited to review the clinic.
    public let invited: Bool
}

public enum SurveyFindingKind: String, Decodable, Sendable, Equatable {
    /// Worse than this patient's own previous answer, by enough to mean something.
    case worsened
    /// Past the question's own threshold, whatever the trend.
    case severe
}

public struct SurveyFinding: Decodable, Sendable, Equatable, Identifiable {
    public let kind: SurveyFindingKind
    public let questionId: String
    public let questionText: String
    public let value: Int
    /// The same question last time. Absent on a `severe` finding.
    public let previous: Int?

    public var id: String { "\(kind.rawValue)-\(questionId)" }

    public var localizedKind: String { L10n.string("survey.finding.\(kind.rawValue)") }
}

public struct SurveyPoint: Decodable, Sendable, Equatable, Identifiable {
    public let assignmentId: String
    public let milestoneDays: Int
    public let submittedAt: Date
    /// Question id to answer.
    public let values: [String: Int]
    public let answeredCount: Int
    public let questionCount: Int
    /**
     * Too little was answered for this point to sit beside a full one.
     *
     * Still drawn — the answers are real — but a chart has to mark it, or a
     * single answer out of five reads as a complete assessment.
     */
    public let partial: Bool

    public var id: String { assignmentId }
}

public struct SurveyTemplateInfo: Decodable, Sendable, Equatable {
    public let code: String
    public let version: Int
    public let title: String
    public let questions: [SurveyQuestion]
}

public struct PatientSurveys: Decodable, Sendable, Equatable {
    public let template: SurveyTemplateInfo
    /// Oldest first.
    public let series: [SurveyPoint]
    /// From the most recent response only.
    public let latestFindings: [SurveyFinding]
    /// False while there is one response: a line needs two points.
    public let hasTrend: Bool

    /// The series for one question, skipping the responses that left it blank.
    public func points(for questionId: String) -> [(point: SurveyPoint, value: Int)] {
        series.compactMap { point in
            point.values[questionId].map { (point, $0) }
        }
    }

    public var needsAttention: Bool { !latestFindings.isEmpty }
}

public struct SurveysAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    /// Questionnaires waiting for me.
    public func mine() async throws -> [PendingSurvey] {
        try await client.send(Endpoint(method: .get, path: "me/surveys"), as: [PendingSurvey].self)
    }

    /// Answers. A value that does not fit its question is refused by the server.
    public func submit(
        _ assignmentId: String,
        answers: [String: SurveyAnswer]
    ) async throws -> SurveySubmitResult {
        try await client.send(
            Endpoint(
                method: .post,
                path: "me/surveys/\(assignmentId)",
                body: try JSONEncoder.klinik.encode(SubmitBody(answers: answers))
            ),
            as: SurveySubmitResult.self
        )
    }

    public func forPatient(_ patientId: String) async throws -> PatientSurveys {
        try await client.send(
            Endpoint(method: .get, path: "patients/\(patientId)/surveys"),
            as: PatientSurveys.self
        )
    }

    private struct SubmitBody: Encodable {
        let answers: [String: SurveyAnswer]
    }
}

/// One answer, in the shape its question expects.
public enum SurveyAnswer: Encodable, Sendable, Equatable {
    case scale(Int)
    case yesNo(Bool)
    case text(String)

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()

        switch self {
        case let .scale(value): try container.encode(value)
        case let .yesNo(value): try container.encode(value)
        case let .text(value): try container.encode(value)
        }
    }
}

File: android/core/network/src/main/kotlin/xyz/klinik/network/SurveysApi.kt
package xyz.klinik.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Patient-reported outcome questionnaires (spec M18, T6.7).
 *
 * The patient's screen and the clinician's screen read the same data and must
 * not say the same things about it. A patient sees the questions and their own
 * answers; they never see a finding, because "your reported pain has worsened"
 * is a clinical reading and this is not the thing that should deliver one.
 */

@Serializable
enum class SurveyAnswerType {
    SCALE_0_10,
    YES_NO,
    TEXT,
}

@Serializable
enum class SurveyDirection {
    /** Pain, swelling: a higher answer is a worse one. */
    @kotlinx.serialization.SerialName("higher-is-worse")
    HIGHER_IS_WORSE,

    /** Sleep, satisfaction. */
    @kotlinx.serialization.SerialName("higher-is-better")
    HIGHER_IS_BETTER,
}

@Serializable
data class SurveyQuestion(
    val id: String,
    val text: String,
    val type: SurveyAnswerType,
    /** Which way is bad. Absent on questions that are not scales. */
    val direction: SurveyDirection? = null,
    val alarmAt: Int? = null,
    val required: Boolean = false,
) {
    /** Whether a chart can plot this one. */
    val isNumeric: Boolean
        get() = type == SurveyAnswerType.SCALE_0_10 || type == SurveyAnswerType.YES_NO
}

@Serializable
data class PendingSurvey(
    val id: String,
    val title: String,
    val description: String? = null,
    /** Days after the operation this one is about. */
    val milestoneDays: Int,
    val scheduledFor: String,
    /** After this it can no longer be answered. */
    val expiresAt: String? = null,
    val questions: List<SurveyQuestion> = emptyList(),
) {
    /**
     * Whether it is still open.
     *
     * A late answer is refused by the server, so the form must not invite one:
     * asking somebody to fill in five questions and then rejecting them is
     * worse than not asking.
     */
    fun isOpen(nowIso: String): Boolean = expiresAt?.let { it > nowIso } ?: true

    val requiredQuestions: List<SurveyQuestion> get() = questions.filter { it.required }
}

/** What the patient is told back. Deliberately not a reading of their answers. */
@Serializable
data class SurveySubmitResult(val invited: Boolean = false)

@Serializable
enum class SurveyFindingKind {
    /** Worse than this patient's own previous answer, by enough to mean something. */
    @kotlinx.serialization.SerialName("worsened")
    WORSENED,

    /** Past the question's own threshold, whatever the trend. */
    @kotlinx.serialization.SerialName("severe")
    SEVERE,
    ;

    val stringKey: String get() = "survey_finding_${name.lowercase()}"
}

@Serializable
data class SurveyFinding(
    val kind: SurveyFindingKind,
    val questionId: String,
    val questionText: String,
    val value: Int,
    /** The same question last time. Absent on a severe finding. */
    val previous: Int? = null,
)

@Serializable
data class SurveyPoint(
    val assignmentId: String,
    val milestoneDays: Int,
    val submittedAt: String,
    /** Question id to answer. */
    val values: Map<String, Int> = emptyMap(),
    val answeredCount: Int = 0,
    val questionCount: Int = 0,
    /**
     * Too little was answered for this point to sit beside a full one.
     *
     * Still drawn — the answers are real — but a chart has to mark it, or a
     * single answer out of five reads as a complete assessment.
     */
    val partial: Boolean = false,
)

@Serializable
data class SurveyTemplateInfo(
    val code: String,
    val version: Int,
    val title: String,
    val questions: List<SurveyQuestion> = emptyList(),
)

@Serializable
data class PatientSurveys(
    val template: SurveyTemplateInfo,
    /** Oldest first. */
    val series: List<SurveyPoint> = emptyList(),
    /** From the most recent response only. */
    val latestFindings: List<SurveyFinding> = emptyList(),
    /** False while there is one response: a line needs two points. */
    val hasTrend: Boolean = false,
) {
    /** The series for one question, skipping the responses that left it blank. */
    fun points(questionId: String): List<Pair<SurveyPoint, Int>> =
        series.mapNotNull { point -> point.values[questionId]?.let { point to it } }

    val needsAttention: Boolean get() = latestFindings.isNotEmpty()
}

/** One answer, in the shape its question expects. */
sealed interface SurveyAnswer {
    data class Scale(val value: Int) : SurveyAnswer
    data class YesNo(val value: Boolean) : SurveyAnswer
    data class Text(val value: String) : SurveyAnswer

    fun toJson(): JsonElement = when (this) {
        is Scale -> JsonPrimitive(value)
        is YesNo -> JsonPrimitive(value)
        is Text -> JsonPrimitive(value)
    }
}

class SurveysApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    /** Questionnaires waiting for me. */
    suspend fun mine(): List<PendingSurvey> =
        decode(client.send(Endpoint(HttpMethod.GET, "me/surveys")))

    /** Answers. A value that does not fit its question is refused by the server. */
    suspend fun submit(assignmentId: String, answers: Map<String, SurveyAnswer>): SurveySubmitResult {
        val body = JsonObject(
            mapOf("answers" to JsonObject(answers.mapValues { (_, answer) -> answer.toJson() })),
        )

        return decode(
            client.send(
                Endpoint(HttpMethod.POST, "me/surveys/$assignmentId", body = body.toString()),
            ),
        )
    }

    suspend fun forPatient(patientId: String): PatientSurveys =
        decode(client.send(Endpoint(HttpMethod.GET, "patients/$patientId/surveys")))

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}
